api/src/cache.ts

```diff
@@ -5,46 +5,85 @@ import { compress, decompress } from './utils/compress';
 import { getConfigFromEnv } from './utils/get-config-from-env';
 import { getMilliseconds } from './utils/get-milliseconds';
 import { validateEnv } from './utils/validate-env';
+import { getMessenger } from './messenger';
+import { getSimpleHash } from '@directus/shared/utils';
+import { SchemaOverview } from '@directus/shared/types';
 
 let cache: Keyv | null = null;
 let systemCache: Keyv | null = null;
+let localSchemaCache: Keyv | null = null;
+let sharedSchemaCache: Keyv | null = null;
 let lockCache: Keyv | null = null;
+let messengerSubscribed = false;
 
-export function getCache(): { cache: Keyv | null; systemCache: Keyv; lockCache: Keyv } {
+type Store = 'memory' | 'redis' | 'memcache';
+
+const messenger = getMessenger();
+
+if (!messengerSubscribed) {
+	messengerSubscribed = true;
+
+	messenger.subscribe('schemaChanged', async (opts) => {
+		if (env.CACHE_STORE === 'memory' && cache && env.CACHE_AUTO_PURGE && opts?.autoPurgeCache !== false) {
```

Comment:
Should we move these conditions up? Since there is actually no reason to subscribe if these do not apply.

```suggestion
if (env.CACHE_STORE === 'memory' && env.CACHE_AUTO_PURGE && !messengerSubscribed) {
	messengerSubscribed = true;

	messenger.subscribe('schemaChanged', async (opts) => {
		if (cache && opts?.autoPurgeCache !== false) {
```